### PR TITLE
Electrified component

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
Adds a component for airlocks and vending machines to use for their electrification behaviour since its essentially a carbon copy for both.
Can be expanded to apply to other types of machines in future.

This has been tested and appears to work.

## Why It's Good For The Game
Lets badmins electrify any machine they want.

## Changelog
:cl:
code: Anything can be electrified now with electrification being a component
/:cl:

### CODE REVIEW

File: code/datums/wires/airlock.dm [hunk 1]
```diff
@@ -18,7 +19,7 @@
 
 /datum/wires/airlock/interactable(mob/user)
 	var/obj/machinery/door/airlock/A = holder
-	if(!issilicon(user) && A.isElectrified() && A.shock(user, 100))
+	if(!issilicon(user) && (SEND_SIGNAL(A, COMSIG_AIRLOCK_WIRES_INTERACT, user) & COMSIG_ELECTRIFIED_SHOCK))
```
Comment (ninjanomnom) [inline code/datums/wires/airlock.dm]: you should not have side effects in condition check functions like this

File: code/datums/components/electrified.dm [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	log_electrification(src, user)
+
+	registersignals()
+
+	if(duration)
+		if(duration_timer && timeleft(duration_timer) >= duration)
+			return
+		duration_timer = addtimer(CALLBACK(src, .proc/end_electrification), duration, TIMER_STOPPABLE|TIMER_OVERRIDE)
+
+	RegisterSignal(src, COMSIG_ELECTRIFICATION_REMOVAL, .proc/end_electrification)
```
Comment (ninjanomnom) [inline code/datums/components/electrified.dm]: This is what GetComponent is for, put removal behaviour in Destroy

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/datums/wires/airlock.dm (hunk 1 above)
Comment (spookydonut) [reply code/datums/wires/airlock.dm]: I'll make a different PR to fix this, how should it be done?

File: code/datums/components/electrified.dm (hunk 2 above)
Comment (spookydonut) [reply code/datums/components/electrified.dm]: It's for the stuff where it toggles electrification, this signal is for that stuff to either <disable the electrification> and get a return value and if not then add electrification

Comment (spookydonut) [reply code/datums/components/electrified.dm]: So should that use like
```
var/a = GetComponent(electrification)
if(a)
    qdel(a)
else
    AddComponent(electrification)
```
?

File: code/datums/wires/airlock.dm (hunk 1 above)
Comment (spookydonut) [reply code/datums/wires/airlock.dm]: Disregard

File: code/datums/components/electrified.dm [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	if(electrification_state == MACHINE_ELECTRIFIED_PERM_TOGGLE)
+		src.electrification_state = MACHINE_ELECTRIFIED_PERMANENT
+	else
+		src.electrification_state = electrification_state
+
+	. = ..()
+
+	ui_update()
+
+	RegisterSignal(parent, COMSIG_MACHINERY_POWERRESTORED, .proc/power_restored)
```
Comment (optimumtact) [inline code/datums/components/electrified.dm]: RegisterWithParent

and UnregisterFromParent

File: code/datums/components/electrified.dm [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	if(duration_timer && electrification_state == MACHINE_ELECTRIFIED_TEMPORARY)
+		clear_timer()
+
+/datum/component/electrified/proc/power_restored(datum/source)
+	registersignals()
+	if(duration && electrification_state == MACHINE_ELECTRIFIED_TEMPORARY)
+		addtimer(CALLBACK(src, .proc/end_electrification), duration, TIMER_STOPPABLE|TIMER_OVERRIDE)
+		duration = null
+
+/datum/component/electrified/proc/registersignals()
```
Comment (optimumtact) [inline code/datums/components/electrified.dm]: RegisterWithParent

and UnregisterFromParent

**Review by optimumtact — CHANGES_REQUESTED** [review]

File: code/__DEFINES/components.dm [hunk 5]
```diff
@@ -213,6 +213,23 @@
...
 #define COMSIG_OBJ_UPDATE_ICON	"obj_update_icon"				//Called by obj to itself whenever update_icon is run. Useful for containers that want to detect this change
+#define COMSIG_OBJ_ATTACK_ANIMAL "obj_attack_animal"			//from base obj/attack_animal()
+
+// /obj/machinery signals
+#define COMSIG_MACHINERY_POWERRESTORED "machinery_powerrestored"	//from base machinery/power_change()
+#define COMSIG_MACHINERY_POWERLOSS "machinery_powerloss"
+#define COMSIG_MACHINERY_BROKEN "machinery_broken"
+
+#define COMSIG_AIRLOCK_BUMPOPEN "airlock_bumpopen"					//from airlock bumpopen()
+	#define COMSIG_ELECTRIFIED_SHOCK 2							// COMPONENT_NO_ATTACK_HAND is 1
```
Comment (optimumtact) [inline code/__DEFINES/components.dm]: something went wrong

File: code/datums/components/electrified.dm (hunk 4 above)
Comment (optimumtact) [reply code/datums/components/electrified.dm]: scrap this, just call these from those procs

File: code/__DEFINES/components.dm [hunk 6]
```diff
@@ -213,6 +213,23 @@
...
 #define COMSIG_OBJ_UPDATE_ICON	"obj_update_icon"				//Called by obj to itself whenever update_icon is run. Useful for containers that want to detect this change
+#define COMSIG_OBJ_ATTACK_ANIMAL "obj_attack_animal"			//from base obj/attack_animal()
+
+// /obj/machinery signals
+#define COMSIG_MACHINERY_POWERRESTORED "machinery_powerrestored"	//from base machinery/power_change()
+#define COMSIG_MACHINERY_POWERLOSS "machinery_powerloss"
+#define COMSIG_MACHINERY_BROKEN "machinery_broken"
+
+#define COMSIG_AIRLOCK_BUMPOPEN "airlock_bumpopen"					//from airlock bumpopen()
+	#define COMSIG_ELECTRIFIED_SHOCK 1
```
Comment (optimumtact) [inline code/__DEFINES/components.dm]: what is going on

Comment (spookydonut) [reply code/__DEFINES/components.dm]: \shrug

Comment (Qustinnus) [reply code/__DEFINES/components.dm]: Why is it a number

Comment (Qustinnus) [reply code/__DEFINES/components.dm]: Nevermind I see

Comment (ninjanomnom) [reply code/__DEFINES/components.dm]: It's a bitflag return value for the above signal, this is the normal format in this file

File: code/__DEFINES/components.dm [hunk 7]
```diff
@@ -213,11 +213,22 @@
...
 #define COMSIG_MACHINERY_BROKEN "machinery_broken"				//from /obj/machinery/obj_break(damage_flag): (damage_flag)
 #define COMSIG_MACHINERY_POWER_LOST "machinery_power_lost"			//from base power_change() when power is lost
 #define COMSIG_MACHINERY_POWER_RESTORED "machinery_power_restored"	//from base power_change() when power is restored
+#define COMSIG_AIRLOCK_BUMPOPEN "airlock_bumpopen"					//from airlock bumpopen()
+	#define COMSIG_ELECTRIFIED_SHOCK 1
+#define COMSIG_AIRLOCK_ATTACKBY_CONDUCTIVE "airlock_attackby_conductive" //from airlock attackby()
+#define COMSIG_AIRLOCK_ATTACKBY_POWERCROWBAR "airlock_attackby_powercrowbar"
+#define COMSIG_AIRLOCK_ATTACK_ALIEN "airlock_attack_alien"
+#define COMSIG_ELECTRIFICATION_CHANGE "electrification_change"
+#define COMSIG_AIRLOCK_AI_ALTCLICK "airlock_ai_altclick"
```
Comment (AnturK) [inline code/__DEFINES/components.dm]: Signal names referencing their singe use are icky. Name this something like AIRLOCK_SILICON_ELECTRIFY

File: code/datums/components/electrified.dm [hunk 8]
```diff
@@ -0,0 +1,279 @@
+GLOBAL_LIST_INIT(unelectrifiable_machines, typecacheof(list(/obj/machinery/door/airlock/cult)))
```
Comment (AnturK) [inline code/datums/components/electrified.dm]: Why does this need to exist ? Wouldn't this make more sense as can electrify signal ?

File: code/__DEFINES/components.dm (hunk 7 above)
Comment (ninjanomnom) [reply code/__DEFINES/components.dm]: You have this backwards, signals should be named after what's happening to cause it, not what the signal should do

File: code/_onclick/ai.dm [hunk 9]
```diff
@@ -147,7 +147,7 @@
 	if(obj_flags & EMAGGED)
 		return
 
-	if(!secondsElectrified)
+	if(!SEND_SIGNAL(src, COMSIG_AIRLOCK_SILICON_ELECTRIFY))
```
Comment (ninjanomnom) [inline code/_onclick/ai.dm]: Flag, not bool

Comment (ninjanomnom) [inline code/_onclick/ai.dm]: This should be moved to the base of AIAltClick and make this call parent

File: code/datums/components/electrified.dm [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+	var/electrification_state = MACHINE_NOT_ELECTRIFIED
+	var/duration
+	var/duration_timer
+
+/datum/component/electrified/Initialize(electrification_state, mob/user, duration)
+	log_electrification(src, user)
+
+	if(electrification_state == MACHINE_NOT_ELECTRIFIED)
+		return INITIALIZE_HINT_QDEL
```
Comment (ninjanomnom) [inline code/datums/components/electrified.dm]:
```suggestion
		return COMPONENT_INCOMPATIBLE
```

File: code/datums/components/electrified.dm [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+		COMSIG_MOVABLE_IMPACT), .proc/cooldown_shock)
+
+/datum/component/electrified/proc/unregister_interactions()
+	UnregisterSignal(parent, list(
+		COMSIG_ATOM_ATTACK_HAND, 
+		COMSIG_ATOM_BUMPED,
+		COMSIG_MOVABLE_IMPACT))
+
+/datum/component/electrified/machinery/register_interactions()
+	return
```
Comment (ninjanomnom) [inline code/datums/components/electrified.dm]: ???

File: code/datums/components/electrified.dm [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+				return
+			src.electrification_state = MACHINE_ELECTRIFIED_EMP
+			addtimer(CALLBACK(src, .proc/end_electrification), MACHINE_DEFAULT_EMP_ELECTRIFY_TIME, TIMER_STOPPABLE|TIMER_OVERRIDE)
+
+/**
+  * Returns something to indicate there's electrification
+  *
+  */
+/datum/component/electrified/machinery/proc/electrified()
+	return 1
```
Comment (ninjanomnom) [inline code/datums/components/electrified.dm]: Use flag defines

File: code/datums/components/electrified.dm [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+		COMSIG_MACHINERY_POWER_LOST,
+		COMSIG_AIRLOCK_SILICON_ELECTRIFY,
+		COMSIG_MACHINERY_BROKEN
+	))
+
+/datum/component/electrified/Destroy(force, silent)
+	unregister_interactions()
+	return ..()
+
+/datum/component/electrified/machinery/Destroy(force, silent)
```
Comment (ninjanomnom) [inline code/datums/components/electrified.dm]: Dont intermix two different types like this

File: code/game/objects/obj_defense.dm [hunk 14]
```diff
@@ -104,6 +104,7 @@
 		playsound(src.loc, 'sound/weapons/slash.ogg', 100, TRUE)
 
 /obj/attack_animal(mob/living/simple_animal/M)
+	SEND_SIGNAL(src, COMSIG_OBJ_ATTACK_ANIMAL, M)
```
Comment (ninjanomnom) [inline code/game/objects/obj_defense.dm]: Put the signal at the base proc and call parent here

File: code/game/machinery/doors/airlock_types.dm [hunk 15]
```diff
@@ -446,17 +446,14 @@
 /obj/machinery/door/airlock/cult/Initialize()
 	. = ..()
 	new openingoverlaytype(loc)
+	RegisterSignal(src, COMSIG_CAN_ELECTRIFY, .proc/dont_electrify)
```
Comment (ninjanomnom) [inline code/game/machinery/doors/airlock_types.dm]: Don't add the component in the first place, this is a poor signal

File: code/game/machinery/doors/airlock.dm [hunk 16]
```diff
@@ -1006,9 +965,8 @@
 		return
 	if(hasPower())
 		if(forced)
-			if(isElectrified())
-				shock(user,100)//it's like sticking a forck in a power socket
-				return
+			if(SEND_SIGNAL(src, COMSIG_AIRLOCK_ATTACKBY_POWERCROWBAR, user) & COMSIG_ELECTRIFIED_SHOCK)
```
Comment (ninjanomnom) [inline code/game/machinery/doors/airlock.dm]: This signal should be sent regardless of state of the airlock, let the component decide if they get shocked

File: code/game/machinery/doors/airlock.dm [hunk 17]
```diff
@@ -799,10 +758,10 @@
 
 
 /obj/machinery/door/airlock/attackby(obj/item/C, mob/user, params)
-	if(!issilicon(user) && !IsAdminGhost(user))
-		if(isElectrified())
-			if(shock(user, 75))
-				return
+	if(prob(75))
+		if(SEND_SIGNAL(src, COMSIG_AIRLOCK_ATTACKBY_CONDUCTIVE, user) & COMSIG_ELECTRIFIED_SHOCK)
```
Comment (ninjanomnom) [inline code/game/machinery/doors/airlock.dm]: 1. The component should be deciding when it shocks
2. The signal should be sent every time if you're going to have this
3. Call parent and use the base attackby signal instead

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by actioninja** [comment]
stale

**Comment by spookydonut** [comment]
> 
> 
> stale

probably need to start over on this

**Pull request closed without merging**